### PR TITLE
Move service strategy to registry/service package

### PR DESCRIPTION
**Pull request by feihujiang**
Ref #12207 Fix todo (// TODO: move to a service specific package. in api/rest/types.go). Move svcStrategy to registry/service package in this PR.

### CODE REVIEW

**Comment by feihujiang** [comment]
@k8s-bot test this please

**Comment by feihujiang** [comment]
I found I missed something. When I finish, I'll reopen this PR.

**Comment by wojtek-t** [comment]
@lavalamp - FYI - I also solved it independently in #16495 - you can choose which one is better :)

File: pkg/api/rest/create_test.go [hunk 1]
```diff
@@ -18,24 +18,8 @@ package rest
 
 import (
 	"testing"
-
-	"k8s.io/kubernetes/pkg/api"
-	"k8s.io/kubernetes/pkg/api/errors"
 )
 
+// TODO: The origin test depends on service strategy, is moved to the registry/service package.
 func TestCheckGeneratedNameError(t *testing.T) {
```
Comment (lavalamp) [inline pkg/api/rest/create_test.go]: Remove empty testing functions

File: pkg/api/rest/update_test.go [hunk 2]
```diff
@@ -18,93 +18,10 @@ package rest
...
-			SessionAffinity: "None",
-			Type:            api.ServiceTypeClusterIP,
-			Ports:           []api.ServicePort{{Name: "p", Protocol: "TCP", Port: 8675, TargetPort: util.NewIntOrStringFromInt(8675)}},
-		},
-	}
-}
-
+// TODO: The origin test depends on service strategy, is moved to the registry/service package.
 // TODO: This should be done on types that are not part of our API
 func TestBeforeUpdate(t *testing.T) {
```
Comment (lavalamp) [inline pkg/api/rest/update_test.go]: Remove empty testing functions

File: pkg/registry/service/strategy.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	return true
+}
+
+func (svcStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) fielderrors.ValidationErrorList {
+	return validation.ValidateServiceUpdate(old.(*api.Service), obj.(*api.Service))
+}
+
+func (svcStrategy) AllowUnconditionalUpdate() bool {
+	return true
+}
```
Comment (lavalamp) [inline pkg/registry/service/strategy.go]: Can you move the matching predicate here, too? As in #16495.

**Comment by feihujiang** [comment]
@k8s-bot test this please

File: pkg/registry/service/strategy.go (hunk 3 above)
Comment (feihujiang) [inline pkg/registry/service/strategy.go]: Yes, I moved too. please review.

**Comment by wojtek-t** [comment]
Thanks - LGTM

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

File: pkg/api/rest/create_test.go [hunk 4]
```diff
@@ -16,26 +16,4 @@ limitations under the License.
...
-	if err := CheckGeneratedNameError(Services, expect, &api.Pod{}); err != expect {
-		t.Errorf("AlreadyExists should be returned when no GenerateName field: %v", err)
-	}
-
-	expect = errors.NewAlreadyExists("foo", "bar")
-	if err := CheckGeneratedNameError(Services, expect, &api.Pod{ObjectMeta: api.ObjectMeta{GenerateName: "foo"}}); err == nil || !errors.IsServerTimeout(err) {
-		t.Errorf("expected try again later error: %v", err)
-	}
-}
+// The original "TestCheckGeneratedNameError" test depends on service strategy, is moved to the registry/service package.
```
Comment (lavalamp) [inline pkg/api/rest/create_test.go]: I think, just remove the whole file :)

File: pkg/api/rest/update_test.go [hunk 5]
```diff
@@ -16,95 +16,4 @@ limitations under the License.
...
-		err := BeforeUpdate(Services, ctx, runtime.Object(&oldSvc), runtime.Object(&newSvc))
-		if tc.expectErr && err == nil {
-			t.Errorf("unexpected non-error for %q", tc.name)
-		}
-		if !tc.expectErr && err != nil {
-			t.Errorf("unexpected error for %q: %v", tc.name, err)
-		}
-	}
-}
+// The original "TestBeforeUpdate" test depends on service strategy, is moved to the registry/service package.
```
Comment (lavalamp) [inline pkg/api/rest/update_test.go]: Yeah, sorry, I didn't realize you removed all the tests from these files-- just remove the file in that case. The git history will tell us what happened.

**Comment by lavalamp** [comment]
Just tiny nits, thanks for cleanup :)

File: pkg/api/rest/create_test.go (hunk 4 above)
Comment (feihujiang) [inline pkg/api/rest/create_test.go]: OK, I removed the test to registry/service/strategy_test.go, and deleted the whole file.

File: pkg/api/rest/update_test.go (hunk 5 above)
Comment (feihujiang) [inline pkg/api/rest/update_test.go]: OK, did the same as the create_test.go .

**Comment by wojtek-t** [comment]
LGTM - thanks for this change!

**Comment by feihujiang** [comment]
@wojtek-t The ```func ObjectMetaFieldsSet(objectMeta api.ObjectMeta) fields.Set``` is changed to
```func ObjectMetaFieldsSet(objectMeta api.ObjectMeta, hasNamespaceField bool) fields.Set```,  so I rebased this commit, please review again, thanks.

**Comment by wojtek-t** [comment]
lgtm - thanks!

**Pull request merged**